pyvista/core/dataobject.py

```diff
@@ -189,39 +189,172 @@ def copy(self, deep=True):
         Parameters
         ----------
         deep : bool, optional
-            When True makes a full copy of the object.
+            When ``True`` makes a full copy of the object.  When
+            ``False``, performs a shallow copy where the points, cell,
+            and data arrays are pointers to the original object.
 
         Returns
         -------
-        newobject : same as input
-           Deep or shallow copy of the input.
+        :class:`pyvista.DataSet`
+            Deep or shallow copy of the input.
+
+        Examples
+        --------
+        Create and make a deep copy of a PolyData object.
+
+        >>> import pyvista
+        >>> mesh_a = pyvista.Sphere()
+        >>> mesh_b = mesh_a.copy()
+        >>> mesh_a == mesh_b
+        True
 
         """
         thistype = type(self)
         newobject = thistype()
+
+        # this must be called to ensure scalars active are "active"
+        # before performing a deep copy
+        self.active_scalars_info
+
         if deep:
             newobject.deep_copy(self)
         else:
             newobject.shallow_copy(self)
         newobject.copy_meta_from(self)
         return newobject
 
+    def __eq__(self, other):
+        """Test equalvency between data objects."""
+        if not isinstance(self, type(other)):
+            return False
+
+        # check if points are equal if DataSet contains points
+        # this also checks if the number of points is equal
+        if hasattr(self, 'points'):
+            if not np.array_equal(self.points, other.points):
+                return False
+
+        # check if cells are equal if DataSet contains cells
+        # this also checks if the number of cells is equal
+        if hasattr(self, 'cells'):
+            if not np.array_equal(self.cells, other.cells):
+                return False        
+
+        # check if cells are equal if DataSet contains cells
+        # this also checks if the number of cells is equal
+        if hasattr(self, 'faces'):
+            if not np.array_equal(self.faces, other.faces):
+                return False        
+
+        if self.field_arrays != other.field_arrays:
+            return False
+
+        if hasattr(self, 'point_arrays'):
+            if self.point_arrays != other.point_arrays:
+                return False
+
+        if hasattr(self, 'cell_arrays'):
+            if self.cell_arrays != other.cell_arrays:
+                return False
```

Comment:
Probably should have kept `__eq__` for a different PR...

> 1. when not hasattr(other, cells) etc. then these checks can still raise

This is addressed by `if not isinstance(self, type(other))`.  If `hasattr(self, 'cells')` then `other` must also have that attribute.

> 2. it seemed weird to only not check the existence of field_arrays, so I tried looking at every concrete subclass of DataSet, but it seemed too much work to get robust results (some concrete subclasses can't be instantiated by just calling the type). In related news, I hope there was a good reason for making our abstract base classes abstract with __new__ hacks rather than some dedicated machinery :P

Issue for a separate PR.  The deeper I go into this code, the more I realize what we need to fix.

> 3. It would probably by safer and cleaner to loop over all these attributes, and check existence on both objects. I guess we'd need two groups for the != and the not np.array_equal kinds.

This might be cleaner.  I'll try an alternative implementation.

> 4. Things like .verts, .lines are derived properties from points and faces, and things like .t_coords, .textures and .vectors are derived from point/cell arrays, right?

Correct.  We can get away with having `t_coords`, `textures`, and `vectors` checked in within the point and cell data.  Would have to check verts and lines here, and maybe (3) will help us with that.

Reply:
Added in da4e20b43238f11ab3ad33fa1b053d5fe92cea95 and 54530cd9f2b74d4b41af5c09c6adf8831c743e3f.